### PR TITLE
Migrate THCState to ATen

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66391
* #66390
* #66389
* #66388
* __->__ #65948

This guts `THCState` to simply be an empty struct, as well as:
- moving `THCState_getPeerToPeerAccess` and its cache into `ATen`.
- cleaning up dead code in `THCGeneral.cpp`
- moving `THCudaInit` and `THCMagma_init` into `CUDAHooks::initCUDA`

Differential Revision: [D31386275](https://our.internmc.facebook.com/intern/diff/D31386275)

### CODE REVIEW

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31386275).

File: aten/src/ATen/cuda/detail/CUDAHooks.cpp [hunk 1]
```diff
@@ -47,26 +48,28 @@ namespace detail {
...
 
-std::function<void(void)> THCMagma_init;
-
 // NB: deleter is dynamic, because we need it to live in a separate
 // compilation unit (alt is to have another method in hooks, but
 // let's not if we don't need to!)
 std::unique_ptr<THCState, void (*)(THCState*)> CUDAHooks::initCUDA() const {
   C10_LOG_API_USAGE_ONCE("aten.init.cuda");
-  THCState* thc_state = THCState_alloc();
+  THCState* thc_state = new THCState();
```
Comment (ngimel) [inline aten/src/ATen/cuda/detail/CUDAHooks.cpp:56]: can you leave a comment here that thc_state is now an empty struct?

**Review by ngimel — APPROVED** [review]

**Comment by suo** [comment]
This appears to have broken master, see: https://hud.pytorch.org/commit/a6774d6e1f3a88f931a3786b7c2617a74bdaa50e. I am reverting.

For some reason, the relevant windows + cuda tests did not run on your PR: https://github.com/pytorch/pytorch/actions/runs/1321834530. cc @zhouzhuojie @seemethere

**Comment by malfet** [comment]
@peterbell10 when re-landing the PR, can you please add `ciflow/cuda` label to make sure that all CUDA tests are being run?

**Comment by peterbell10** [comment]
Yup, already seeing the test fail again in #66480.

**Pull request closed without merging**